CIPs/cip-0021.md

```diff
@@ -96,15 +97,80 @@ def get_lookback_window(chain: ChainState) -> int:
 
+### Solidity Implementation
+
+```solidity
+
+  struct LookbackWindow {
+    // Value for lookbackWindow before `nextValueActivationBlock`
+    uint256 oldValue;
+    // Value for lookbackWindow after `nextValueActivationBlock`
+    uint256 nextValue;
+    // Epoch where next value is activated
+    uint256 nextValueActivationEpoch;
+  }
+  
+  LookbackWindow public uptimeLookbackWindow;
+
+  event UptimeLookbackWindowSet(uint256 window, uint256 activationEpoch);
+  
+  /**
+   * @notice Sets the uptime lookback window.
+   * @param window New window.
+   */
+  function setUptimeLookbackWindow(uint256 window) public onlyOwner {
+    require(window >= 3 && window <= 720, "UptimeLookbackWindow must be within safe range");
+    require(
+      window <= getEpochSize().sub(2),
+      "UptimeLookbackWindow must be smaller or equal to epochSize - 2"
+    );
+
+    uptimeLookbackWindow.oldValue = _getUptimeLookbackWindow();
+
+    // changes only take place on the next epoch
+    uptimeLookbackWindow.nextValueActivationEpoch = getEpochNumber().add(1);
+    uptimeLookbackWindow.nextValue = window;
+
+    emit UptimeLookbackWindowSet(window, uptimeLookbackWindow.nextValueActivationEpoch);
+  }
+
+  /**
+   * @notice Gets the uptime lookback window.
+   */
+  function getUptimeLookbackWindow() public view returns (uint256 lookbackWindow) {
+    lookbackWindow = _getUptimeLookbackWindow();
+    require(lookbackWindow != 0, "UptimeLookbackWindow is not initialized");
+  }
+
+  /**
+   * @notice Gets the uptime lookback window.
+   */
+  function _getUptimeLookbackWindow() internal view returns (uint256 lookbackWindow) {
+    if (getEpochNumber() >= uptimeLookbackWindow.nextValueActivationEpoch) {
+      return uptimeLookbackWindow.nextValue;
+    } else {
+      return uptimeLookbackWindow.oldValue;
+    }
+  }
+```
+
+### Scenario Considerations
+
+In the current specification, the following migration scenarios are considered:
+
+ 1. **Hardfork is activated before Smart Contract is upgraded:** On this scenario, on the first epoch after hardfork the blockchain client will fail to obtain `uptimeLookbackWindow` from the contract; and will use `DEFAULT_VALUE`. Then, when smart contract is deployed (mid-epoch), it will continue to fail until next epoch where initialized value will be used. Thus, the value remains constant during the epoch.
+ 2. **Smart Contract is upgraded before Hardfork is activated:** On this scenario, on the first epoch after hardfork the blockchain client will obtain `uptimeLookbackWindow` successfully from the smart contract, and use the same value during the epoch.
+ 3. **setUptimeLookbackWindow() is called mid-epoch:** On this scenario, the new value will only be used after epoch change, thus value will remain constant during the epoch.
```

Comment:
```suggestion
 1. **The first post-fork epoch starts before a lookback window from the smart contract is available:** This scenario includes several cases: the smart contract not having been upgraded yet, the smart contract having been upgraded but `setUptimeLookbackWindow()` not having been called, or `setUptimeLookbackWindow()` having been called but its value not having come into effect yet.  In all these cases, in the first epoch after the hard fork, the blockchain client attempt to obtain `uptimeLookbackWindow` from the contract, but will get an error, and will then use `DEFAULT_VALUE`. This will continue until the first block of the first epoch after the contract has been upgraded and `setUptimeLookbackWindow()` has been called successfully, from which point onwards the smart contract call will succeed and return the lookback window which the client will then use. Thus, the value does not change within an epoch.
 2. **The smart Contract is upgraded and initialized before, or in the same epoch as, the hard fork activation:** In this scenario, in the first post-fork epoch, the blockchain client will obtain `uptimeLookbackWindow` successfully from the smart contract, and use that value during the epoch.
 3. **setUptimeLookbackWindow() is called mid-epoch:** On this scenario, the new value come into effect at the beginning of the next epoch, thus the value returned by the contract will remain constant during the epoch.
```

Making this more complicated, but more correct.  There be better ways to phrase things, though.